crates/polars-parquet/src/arrow/write/nested/dremel/mod.rs

```diff
@@ -0,0 +1,422 @@
+//! Implements the Dremel encoding part of Parquet with *repetition-levels* and *definition-levels*
+
+use arrow::bitmap::Bitmap;
+use arrow::offset::OffsetsBuffer;
+use polars_utils::fixedringbuffer::FixedRingBuffer;
+
+use super::super::pages::Nested;
+
+#[cfg(test)]
+mod tests;
+
+/// A Dremel encoding value
+#[derive(Clone, Copy)]
+pub struct DremelValue {
+    /// A *repetition-level* value
+    pub rep: u16,
+    /// A *definition-level* value
+    pub def: u16,
+}
+
+/// This tries to mirror the Parquet Schema structures, so that is simple to reason about the
+/// Dremel structures.
+enum LevelContent<'a> {
+    /// Always 1 instance
+    Required,
+    /// Zero or more instances
+    Repeated,
+    /// Zero or one instance
+    Optional(Option<&'a Bitmap>),
+}
+
+struct Level<'a> {
+    content: LevelContent<'a>,
+    /// "Iterator" with number of elements for the next level
+    lengths: LevelLength<'a>,
+    /// Remaining number of elements to process. NOTE: This is **not** equal to `length - offset`.
+    remaining: usize,
+    /// Offset into level elements
+    offset: usize,
+    /// The definition-level associated with this level
+    definition_depth: u16,
+    /// The repetition-level associated with this level
+    repetition_depth: u16,
+}
+
+/// This contains the number of elements on the next level for each
+enum LevelLength<'a> {
+    /// Fixed number of elements based on the validity of this element
+    Optional(usize),
+    /// Fixed number of elements irregardless of the validity of this element
+    Constant(usize),
+    /// Variable number of elements and calculated from the difference between two `i32` offsets
+    OffsetsI32(&'a OffsetsBuffer<i32>),
```

Comment:
Not for now, but I think we can remove this. We should only support the larger index types.